nucypher/policy/conditions/evm.py

```diff
@@ -168,41 +169,64 @@ def _next_endpoint(
+            self._check_chain_id(w3)
+        return self._execute(w3, resolved_parameters)
+
     def execute(self, providers: Dict[int, Set[HTTPProvider]], **context) -> Any:
         resolved_parameters = []
         if self.parameters:
             resolved_parameters = resolve_any_context_variables(
                 self.parameters, **context
             )
 
-        endpoints = self._next_endpoint(providers=providers)
-        latest_error = ""
-        for provider in endpoints:
-            w3 = self._configure_provider(provider)
-            try:
-                result = self._execute(w3, resolved_parameters)
-                break
-            except RequiredContextVariable:
-                raise
-            except Exception as e:
-                latest_error = f"RPC call '{self.method}' failed: {e}"
-                self.LOG.warn(f"{latest_error}, attempting to try next endpoint.")
-                # Something went wrong. Try the next endpoint.
-                continue
-        else:
-            # Fuck.
+        # use local rpc endpoint if available
+        if self.chain in providers:
+            latest_error = ""
+            for provider in self._next_endpoint(providers=providers):
+                result, error = self._execute_with_provider(
+                    provider, resolved_parameters
+                )
+                if result:
+                    return result
+                latest_error = error
+
             raise RPCExecutionFailed(
                 f"RPC call '{self.method}' failed; latest error - {latest_error}"
             )
 
-        return result
+        if self.rpc_endpoint:
+            result, error = self._execute_with_provider(
+                HTTPProvider(self.rpc_endpoint), resolved_parameters, check_chain=False
+            )
+            if error:
+                raise RPCExecutionFailed(error)
+            return result
 
-    def _execute(self, w3: Web3, resolved_parameters: List[Any]) -> Any:
-        """Execute onchain read and return result."""
-        rpc_endpoint_, rpc_method = self.method.split("_", 1)
-        rpc_function = self._get_web3_py_function(w3, rpc_method)
-        rpc_result = rpc_function(*resolved_parameters)  # RPC read
-        return rpc_result
+        raise NoConnectionToChain(chain=self.chain)
+
+    def _execute(
+        self, w3: Web3, resolved_parameters: List[Any]
+    ) -> Tuple[Optional[Any], Optional[str]]:
+        """Execute onchain read and return result and error if any."""
+        try:
+            rpc_endpoint_, rpc_method = self.method.split("_", 1)
+            rpc_function = self._get_web3_py_function(w3, rpc_method)
+            rpc_result = rpc_function(*resolved_parameters)  # RPC read
+            return rpc_result, None
+        except RequiredContextVariable:
+            raise
+        except Exception as e:
+            error = f"RPC call '{self.method}' failed: {e}"
+            self.LOG.warn(error)
+            return None, error
```

Comment:
How come you decided to move this here, and not continue to allow this method raise the exception, and it be handled at the top level? Maybe I'm missing something, but `_execute` should propagate the exception, and let the code at the higher level handle it appropriately. It feels like the use of multiple endpoints is now creeping into a specific method for execution a rpc call, and any other methods in the call stack before the code to try a different endpoint.

Same comment in other spots (contract, time) where the same thing is done.

Reply:
at the time i thought i was removing duplicated code, but i hadn't realised that i'd need to repeat this for several conditions. You're right, it's a downgrade - I've reset the behaviour